codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpProtocolGeneratorUtils.java

```diff
@@ -106,4 +106,139 @@ private HttpProtocolGeneratorUtils() {}
 
         return errorShapes;
     }
+
+    /**
+     * Generates a function that handles error deserialization by getting the error code then
+     * dispatching to the error-specific deserializer.
+     *
+     * If the error code does not map to a known error, a generic error will be returned using
+     * the error code and error message discovered in the response.
+     *
+     * The default error message and code are both "UnknownError".
+     *
+     * @param context The generation context.
+     * @param operation The operation to generate for.
+     * @param responseType The response type for the HTTP protocol.
+     * @param errorMessageCodeGenerator A consumer that generates a snippet that sets the {@code errorCode}
+     *                                  and {@code errorMessage} variables from the http response.
+     * @return A set of all error structure shapes for the operation that were dispatched to.
+     */
+    public static Set<StructureShape> generateXmlErrorDispatcher(
```

Comment:
I don't think this is necessary. The only meaningful difference between this and what already exists is that it creates a decoder and does error wrapping, both of which can be done by the error decode function.

Reply:
I think separating them there, reduces the additional code required in error decode function which has to be implemented by other protocols. 

Also wanted to keep error dispatcher separate to better accommodate customization's around error deserialization. If this is not used, we can come back later and revert this. Can add a TODO instead.

Reply:
In what way is it reduced? You can always have a stub function to generate an XML deserializer that's shared - you do have that function in fact, you would just want to move it to where you actually need it. The problem now is that you have a lot of duplicated code to handle generic dispatch. That and there are abstract functions that are never guaranteed to be called now.

Is this change because you want to use this to handle S3's 200 errors? I was under the impression that there was another strategy for that just using a middleware.

Additionally, I'm not sure the XML specific bits should be public (at least for now). I would move the XML stuff over to the sdk repo.